src/qtcwin.c

```diff
@@ -0,0 +1,1814 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+	/* ------------------------------------------------------------
+	 *        Handle QTC panel, receive and send QTC, write to log
+	 *
+	 *--------------------------------------------------------------*/
+
+#include "qtcwin.h"
+
+#include "onechar.h"
+#include <panel.h>
+#include "nicebox.h"
+#include "time_update.h"
+#include <ctype.h>
+#include "rtty.h"
+#include "sendbuf.h"
+#include "log_recv_qtc_to_disk.h"
+#include "log_sent_qtc_to_disk.h"
+#include "globalvars.h"
+#include "qtcutil.h"
+#include "genqtclist.h"
+#include "speedupndown.h"
+#include "cw_utils.h"
+#include "keyer.h"
+#include "callinput.h"
+#include "get_time.h"
+#include "write_keyer.h"
+
+#include <sys/time.h>
+
+#define DIRCLAUSE (direction == RECV) || (direction == SEND && (activefield == 0 || activefield == 2))
+
+extern char hiscall[];
+extern char lastcall[];
+extern int trxmode;
+extern t_qtcreclist qtcreclist;
+extern t_qtclist qtclist;
+extern int keyerport;
+extern int nr_qsos;
+extern char qtc_recv_msgs[12][80];
+extern char qtc_send_msgs[12][80];
+extern char qtc_phrecv_message[14][80];
+extern char qtc_phsend_message[14][80];
+extern int data_ready;
+extern struct t_qtc_store_obj *qtc_temp_obj;
+extern int qtcrec_record;
+extern char qtcrec_record_command[2][50];
+extern char qtcrec_record_command_shutdown[50];
+extern char wkeyerbuffer[];
+extern t_qtc_ry_line qtc_ry_lines[QTC_RY_LINE_NR];
+extern int qtc_ry_currline;
+extern int qtc_ry_capture;
+
+enum {
+  QTCRECVWINBG = 32,
+  QTCRECVLINE,
+  QTCRECVINVLINE,
+  QTCRECVCURRLINE,
+  QTCRECVBG,
+  KEYERLINE
+};
+
+char sentqtc_queue[11][30];
+int sentqtc_p;
+
+int qtcpanel = 0;
+int currrecstate;
+WINDOW * qtcwin;
+PANEL * qtc_panel;
+WINDOW * ry_win;
+PANEL * ry_panel;
+WINDOW * ry_help_win;
+PANEL * ry_help_panel;
+int activefield;
+// array values: hor position, cursor position, field len to fill with spaces
+int pos[6][3] = {{3, 3, 15}, {3, 6, 4}, {8, 8, 2}, {3, 3, 4}, {8, 8, 14}, {24, 24, 4}};
+int curpos = 0;
+int curfieldlen = 0;
+static char last_rtty_line[2][50] = {"", ""};	// local copy and store to remain
+static char prevqtccall[15] = "";
+int curr_rtty_line = 0;
+int capturing = 0;
+char help_rec_msgs[6][26] = {
+    "Enter callsign",
+    "Enter the QTC serial",
+    "Enter the QTC number",
+    "Enter the time",
+    "Enter the CALL",
+    "Enter the SERIAL"
+};
+char help_send_msgs[7][26] = {
+    "Enter callsign",
+    "",
+    "Enter the QTC number",
+    "Press ENTER to send QTC",
+    "",
+    "",
+    "Press CTRL+S to SAVE!"
+};
+
+char * qtccallsign;
+int * qtccount;
+int * qtccurrdiretion;
+
+int qtc_main_panel(int direction) {
```

Comment:
That is a very big function. It will make it hard to maintain it in the future. I suggest to split it into parts, e.g. for creating the panel, running the loop and for destroying the panel. What do you think